### PR TITLE
[change] Separate action outputs into OutputDistributions object

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Rather than have the output distribution built into the Policy, we separate these out into a separate class. Child classes of `OutputDistribution` include both continuous and discrete versions. This opens the door to adding more types of output distributions in the future without major changes to Policy code. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [x] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/common/nn_policy.py [hunk 1]
```diff
@@ -199,82 +203,33 @@ def _create_cc_actor(
...
-        self.all_log_probs = tf.identity(all_probs, name="action_probs")
-
-        single_dim_entropy = 0.5 * tf.reduce_mean(
-            tf.log(2 * np.pi * np.e) + 2 * log_sigma
+        self.all_log_probs = tf.identity(
+            distribution.total_log_probs, name="action_probs"
         )
-        # Make entropy the right shape
-        self.entropy = tf.ones_like(tf.reshape(mu[:, 0], [-1])) * single_dim_entropy
+        self.entropy = distribution.entropy
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/common/nn_policy.py]: can we aggregate all occurrences of this line i.e. line 283

File: ml-agents/mlagents/trainers/common/distributions.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    @abc.abstractproperty
+    def log_probs(self) -> tf.Tensor:
+        """
+        Returns a Tensor that when evaluated, produces the per-action log probabilities of this distribution.
+        The shape of this Tensor should be equivalent to (batch_size x the number of actions) produced in sample.
+        """
+        pass
+
+    @abc.abstractproperty
+    def total_log_probs(self) -> tf.Tensor:
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/common/distributions.py]: Pretty much - are we 100% sure that reduce_sum would work for all distribution types?

File: ml-agents/mlagents/trainers/common/nn_policy.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/common/nn_policy.py]: 👍 will do it in a follow-up PR. TODO: broader discussion about standardization of discrete and continuous (requires inference changes on C#)

File: ml-agents/mlagents/trainers/common/distributions.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+    @property
+    def sample(self) -> tf.Tensor:
+        return self._sampled_policy
+
+    @property
+    def entropy(self) -> tf.Tensor:
+        return self._entropy
+
+
+class TanhSquashedGaussianDistribution(GaussianDistribution):
```
[the remark this answers is not in the document]
Comment (ervteng) [reply ml-agents/mlagents/trainers/common/distributions.py]: Seems like this distribution is different enough to warrant a separate one - wanted to keep the different flags in each distribution minimal. Am open to making it a flag as well

**Pull request merged**